### DISCUSSION TITLE
Setting default/fallback ControlTheme

### DISCUSSION THREAD

**Discussion by vpenades** [AvaloniaUI/Avalonia#17612] [General]
I've read many tutorials and examples on how to configure a ControlTheme and almost all I've read follow these steps:

- Create a ResourceDictionary with your ControlTheme inside
- Declare the ResourceDictionary within App.axaml

So whenever you want to change the themes, you switch the files in App.axaml.

The problem is that I want to set a _default_ or _fallback_ ControlTheme for every control, so when no theme is chosen in the App.axaml, the control has a theme to fall back to.

The closest I've came to do this is by doing this in the static constructor of the control:

```c#

static CustomControl()
{
    
    var uri = new Uri("avares://AvaloniaControlsLibrary/CustomControl.axaml");
    var include = new ResourceInclude(uri) { Source = uri };
    var resourceDictionary = new ResourceDictionary();
    resourceDictionary.MergedDictionaries.Add(include);

    if (resourceDictionary .TryGetResource(typeof(CustomControl), null, out var res) && res is ControlTheme ctrlTheme)
    {
        ThemeProperty.OverrideDefaultValue<CustomControl>(ctrlTheme);
    }
 }

```

This code successfully sets the default control theme, not as a fallback but as top level theme, so It cannot be easily overriden downstream by setting another controlTheme at app.axaml

What I would like to do is:
- to assign a default ControlTheme for each control so when the renderer founds nothing else to render with, it uses it as a fallback
- I want to avoid having to require adding a resource dictionary at app.axaml to keep the library easy to use.
- I want to be able to override that default controlthem by defining another controltheme downstream, on another UserControl or at app.axaml

I have been expending several days on this, and I'm out of ideas

**Comment by timunie** [comment]
I personally have one ResourceDict in the controls lib that bundles all my custom ones via ResourceInclude. And in App.axaml I only have to add one single line 🤗.

**Reply by vpenades** [reply]
yes it's almost like creating a theme.... yet, I would like to know if it's possible to avoid even that, to just drop the library in a project and begin pulling controls here and there, yet keeping the controls thematizable

**Comment by maxkatz6** [comment]
If you want to force default ControlTheme, without any resources being referenced from the App.xaml.cs, you can do that.

`ThemeProperty` is a normal StyledProperty. Meaning you can override its default value.
`Control.ThemeProperty.OverrideDefaultValue<CustomControl>(new ControlTheme())` from static constructor.
I haven't tried that code, but it should work without hidden issues.

> var uri = new Uri("avares://AvaloniaControlsLibrary/CustomControl.axaml");

I would recommend adding `x:Class` to your `CustomControl.axaml` file, and reference it using C# class name. This helps with avoiding reflection and trimming issues.

**Comment by maxkatz6** [comment]
> The closest I've came to do this is by doing this in the static constructor of the control:

Why did you have to add ResourceDictionary in between? You can use ResourceInclude.Loaded to get content with reflection. Or even better x:Class instead.

**Comment by vpenades** [comment]
@maxkatz6 in either case, although loading the theme to ThemeProperty.OverrideDefaultValue works well in setting the theme to the control, it does not act as a fallback but as a top level value.

What I mean with this is to also be able to define _another_ theme for the control at app.xaml  that overrides the default one defined by the control. And that's not happening.

If I define a default "basic" theme to the control using ThemeProperty.OverrideDefaultValue , and then I define an "advanced" theme of that same control at app.xaml,  the control stays with the "basic" theme

**Comment by vpenades** [comment]
Okey... after some tinkering, I've ended up with this solution:

##  Creating a control and it's default theme:

```xml
<!-- CustomControlDefaultTheme.axaml -->
<ControlTheme x:Class="SomeNamespace.CustomControlDefaultTheme"
              xmlns="https://github.com/avaloniaui"
              xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
              xmlns:controls="using:SomeNamespace.CustomControl"
              TargetType="controls:GroupBox">

  <Setter Property="Margin" Value="4" />

</ControlTheme>
```
```c#
// CustomControl.cs
namespace SomeNamespace;

partial class CustomControlDefaultTheme : ControlTheme {}

class CustomControl : TemplatedControl
{
    static CustomControl()
    {
         // sets CustomControlDefaultTheme as the default theme of CustomControl
         StyledElement.ThemeProperty.OverrideDefaultValue<CustomControl>(new CustomControlDefaultTheme());
    }
}
```

## Overriding default theme

If at a later stage we want to override or replace the default theme, things get tricky:

### what does NOT work:

```xml
<UserControl>
  <CustomControl>
    <CustomControl.Resources>
         <ControlTheme x:Key={x:Type CustomControl} x:DataType="CustomControl">
             ....
         </ControlTheme>
    </CustomControl.Resources>
  </CustomControl>
</UserControl>
```

Actually, defining `<ControlTheme x:Key={x:Type CustomControl} ... />` anywere, in places where it's supposed to override the default theme, like in app.xaml , _does not take any effect_.

### what DOES work:

```xml
<UserControl>
  <CustomControl>
    <CustomControl.Theme>
         <ControlTheme x:DataType="CustomControl">
             ....
         </ControlTheme>
    </CustomControl.Theme>
  </CustomControl>
</UserControl>
```

But this can only be done within an instance of the control itself.

Essentially, you loose the "cascade" effect of Resource Dictionaries.

### My conclusions

Since overriding the default theme apparently does not work if we simply set the new themes in resource dictionaries, the old fashioned
way of creating custom themes is not appropiate anymore.

When working with ControlThemes, I don't know if this approach is appropiate. If it is, which would be the right approach to define additional overriding themes? given that resourcedictionaries can't override the theme's default value?

I think there's a need for an updated, end to end tutorial of good practices to deal with ControlThemes.

**Comment by tomlm** [comment]
What seemed to work for me was to define a selector in my theme file which applied it to descendants like this:
_SampleTheme.xaml_
```xaml
<Styles xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:local="using:Sample"
        x:Class="Sample.MyControls">

  <Styles.Resources>
    <ResourceDictionary>
      <ResourceDictionary.MergedDictionaries>
        <ResourceInclude Source="avares://Sample/Control1.axaml" />
        </ResourceDictionary.MergedDictionaries>
    </ResourceDictionary>
  </Styles.Resources>
  
  <Style Selector=":is(local|Control1)">
    <Setter Property="Theme" Value="{StaticResource {x:Type local:Control1}}"/>
  </Style>
</Styles>
```

And then in 
_App.xaml_
```xaml
  <Application.Styles>
    <FluentTheme />
    <local:SampleTheme/>
  </Application.Styles>
```
If someone sets the theme in the instance file it will override the rule and if anyone adds a selector after mine it should override the rule too.

**Reply by vpenades** [reply]
yes, my point was to do something that worked _without needing to register something at the application.styles level_

In WPF we had something that worked, but I can't quite figure out how to do it in Avalonia.

**Comment by vpenades** [comment]
Elavorating on my question... this is related to creating a TemplatedControl (using the Avalonia's templates)

In WPF, when you created a templated control, it did register something in the static constructor, so the control could be used straight away.

In avalonia, a TemplatedControl created by default simply creates a ResourceDictionary and a code behind class, but nothing is connecting them, so if you try to use the control straight away, nothing is shown.... until you include its associated resourcedictionary into the app.axaml

This is not straightforward and imho quite cumbersome, specially when you're creating custom templatedcontrols all the time.

I believe TemplatedControl classes (specially the ones created using the avalonia design templates) should include a static constructor that registers the resource dictionary, so you don't need to do anything at the app.xaml level
